core/src/main/java/org/elasticsearch/index/engine/ElasticsearchConcurrentMergeScheduler.java

```diff
@@ -174,4 +147,42 @@ public MergeScheduler clone() {
         // the clone will just be the identity.
         return this;
     }
+
+    @Override
+    protected boolean maybeStall(IndexWriter writer) {
+        // Don't stall here, because we do our own index throttling (in InternalEngine.IndexThrottle) when merges can't keep up
+        return true;
+    }
+
+    @Override
+    protected MergeThread getMergeThread(IndexWriter writer, MergePolicy.OneMerge merge) throws IOException {
+        MergeThread thread = super.getMergeThread(writer, merge);
+        thread.setName(EsExecutors.threadName(indexSettings, "[" + shardId.index().name() + "][" + shardId.id() + "]: " + thread.getName()));
+        return thread;
+    }
+
+    MergeStats stats() {
+        MergeStats mergeStats = new MergeStats();
+        // TODO: why would there be more than one CMS for a single shard...?
+        mergeStats.add(totalMerges.count(), totalMerges.sum(), totalMergesNumDocs.count(), totalMergesSizeInBytes.count(),
+                currentMerges.count(), currentMergesNumDocs.count(), currentMergesSizeInBytes.count(),
+                totalMergeStoppedTime.count(),
+                totalMergeThrottledTime.count(),
+                config.isAutoThrottle() ? getIORateLimitMBPerSec() : Double.POSITIVE_INFINITY);
+        return mergeStats;
+    }
+
+    void refreshConfig() {
+        if (this.getMaxMergeCount() != config.getMaxMergeCount() + 1 || this.getMaxThreadCount() != config.getMaxThreadCount()) {
+            // NOTE: we pass maxMergeCount+1 here so that CMS will allow one too many merges to kick off which then allows
+            // InternalEngine.IndexThrottle to detect too-many-merges and throttle:
+            this.setMaxMergesAndThreads(config.getMaxMergeCount() + 1, config.getMaxThreadCount());
+        }
+        if (config.isAutoThrottle() && getIORateLimitMBPerSec() == Double.POSITIVE_INFINITY) {
```

Comment:
Sigh, I'll add that too.

But as this change stands, if the config wants to enable autoIOThrottle but the current CMS has it enabled already, we will then disable it?